Libplanet/Blockchain/Policies/VolatileStagePolicy.cs

```diff
@@ -36,184 +36,203 @@ public VolatileStagePolicy()
             _logger = Log.ForContext<VolatileStagePolicy<T>>();
             Lifetime = lifetime;
-            _set = new ConcurrentDictionary<TxId, Transaction<T>?>();
-            _queue = new List<TxId>();
+            _staged = new ConcurrentDictionary<TxId, Transaction<T>>();
+            _ignored = new ConcurrentBag<TxId>();
             _lock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
         }
 
         /// <summary>
-        /// Volatilizes staged transactions older than this <see cref="Lifetime"/>.
-        /// <para>Note that transactions older than the lifetime never cannot be staged.</para>
+        /// Lifespan for <see cref="Transaction{T}"/>s.  Any <see cref="Transaction{T}"/> older
+        /// than this <see cref="TimeSpan"/> will be considered expired.
         /// </summary>
+        /// <remarks>
+        /// Expired <see cref="Transaction{T}"/>s cannot be staged.
+        /// </remarks>
         public TimeSpan Lifetime { get; }
 
-        /// <inheritdoc cref="IStagePolicy{T}.Stage(BlockChain{T}, Transaction{T})"/>
+        /// <inheritdoc/>
         public void Stage(BlockChain<T> blockChain, Transaction<T> transaction)
         {
-            if (DateTimeOffset.UtcNow - Lifetime > transaction.Timestamp)
+            if (Exipred(transaction))
             {
-                // The transaction is already expired; don't stage it at all.
                 return;
             }
 
             try
             {
-                if (!_set.TryAdd(transaction.Id, transaction))
+                _lock.EnterUpgradeableReadLock();
+                if (_ignored.Contains(transaction.Id))
                 {
-                    _lock.EnterUpgradeableReadLock();
-                    if (_queue.Contains(transaction.Id))
-                    {
-                        return;
-                    }
+                    return;
                 }
-
-                _lock.EnterWriteLock();
-                try
+                else
                 {
-                    _queue.Add(transaction.Id);
-                    const string TimestampFormat = "yyyy-MM-ddTHH:mm:ss.ffffffZ";
-                    _logger
-                        .ForContext("Tag", "Metric")
-                        .Debug(
-                            "Transaction {TxId} by {Signer} " +
-                            "with timestamp {TxTimestamp} staged at {StagedTimestamp}.",
-                            transaction.Id,
-                            transaction.Signer,
-                            transaction.Timestamp.ToString(
-                                TimestampFormat, CultureInfo.InvariantCulture),
-                            DateTimeOffset.UtcNow.ToString(
-                                TimestampFormat, CultureInfo.InvariantCulture));
-                }
-                finally
-                {
-                    _lock.ExitWriteLock();
+                    try
+                    {
+                        _lock.EnterWriteLock();
```

Comment:
According to [MSDN ReadWriteLockSlim](https://docs.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim?redirectedfrom=MSDN&view=net-6.0) example, `ReadWriteLockSlim.EnterWriteLock()` and all kind of lock should be call on outside of `try` block. If you worried about throw error from `ReadWriteLockSlim.EnterWriteLock()`, it is ok. ([ref](https://stackoverflow.com/a/24533469))

Reply:
Thanks, now that I have given more thought to it and going over your reference, it *should* be outside `try`. I'll fix it right away.